--- utils/h2ph.PL
+++ utils/h2ph.PL
@@ -354,6 +354,7 @@
 		use re "eval";
 		my $typelist = join '|', keys %isatype;
 		$new =~ s['
+		  (?:(?:__)?const(?:__)?\s+)?
 		  (?:(?:un)?signed\s+)?
 		  (?:long\s+)?
 		  (?:$typelist)\s+
@@ -362,6 +363,7 @@
 		  ']
 		 [my \$$1]gx;
 		$new =~ s['
+		  (?:(?:__)?const(?:__)?\s+)?
 		  (?:(?:un)?signed\s+)?
 		  (?:long\s+)?
 		  (?:$typelist)\s+
@@ -734,9 +736,15 @@
 # non-GCC?) C compilers, but gcc uses an additional include directory.
 sub inc_dirs
 {
-    my $from_gcc    = `$Config{cc} -v 2>&1`;
-    $from_gcc       =~ s:^Reading specs from (.*?)/specs\b.*:$1/include:s;
-
+    my $from_gcc    = `LC_ALL=C $Config{cc} -v 2>&1`;
+    if( !( $from_gcc =~ s:^Reading specs from (.*?)/specs\b.*:$1/include:s ) )
+    { # gcc-4+ :
+       $from_gcc   = `LC_ALL=C $Config{cc} -print-search-dirs 2>&1`;
+       if ( !($from_gcc =~ s/^install:\s*([^\s]+[^\s\/])([\s\/]*).*$/$1\/include/s) )
+       {
+           $from_gcc = '';
+       };
+    };
     length($from_gcc) ? ($from_gcc, $Config{usrinc}) : ($Config{usrinc});
 }
 
